lib/spack/spack/util/module_cmd.py

```diff
@@ -21,65 +22,97 @@
-        # For python, we use the same python running the Spack process, because
-        # we can guarantee its existence. We have to do some LD_LIBRARY_PATH
-        # shenanigans to ensure python will run.
-
-        # LD_LIBRARY_PATH under which Spack ran
-        os.environ['SPACK_LD_LIBRARY_PATH'] = spack.main.spack_ld_library_path
-
-        # suppress output from module function
-        module_cmd += ' >/dev/null;'
-
-        # Capture the new LD_LIBRARY_PATH after `module` was run
-        module_cmd += 'export SPACK_NEW_LD_LIBRARY_PATH="$LD_LIBRARY_PATH";'
-
-        # Set LD_LIBRARY_PATH to value at Spack startup time to ensure that
-        # python executable finds its libraries
-        module_cmd += 'LD_LIBRARY_PATH="$SPACK_LD_LIBRARY_PATH" '
-
-        # Execute the python command
-        module_cmd += '%s -E -c "%s";' % (sys.executable, py_cmd)
-
-        # If LD_LIBRARY_PATH was set after `module`, dump the old value because
-        # we have since corrupted it to ensure python would run.
-        # dump SPACKIGNORE as a placeholder for parsing if LD_LIBRARY_PATH null
-        module_cmd += 'echo "${SPACK_NEW_LD_LIBRARY_PATH:-SPACKIGNORE}"'
+def module(*args, **kwargs):
+    module_cmd = kwargs.get('module_template', 'module ' + ' '.join(args))
 
-        module_p  = subprocess.Popen(module_cmd,
-                                     stdout=subprocess.PIPE,
-                                     stderr=subprocess.STDOUT,
-                                     shell=True,
-                                     executable="/bin/bash")
-
-        # Cray modules spit out warnings that we cannot supress.
-        # This hack skips to the last output (the environment)
-        env_out = str(module_p.communicate()[0].decode()).strip().split('\n')
-
-        # The environment dumped as json
-        env_json = env_out[-2]
-        # Either the uncorrupted $LD_LIBRARY_PATH or SPACKIGNORE
-        new_ld_library_path = env_out[-1]
+    if args[0] in module_change_commands:
+        use_env_null = platform.system().lower() == 'linux'
+        if use_env_null:
+            module_cmd += ' >/dev/null 2>&1; /usr/bin/env -0'
+
+            module_p  = subprocess.Popen(module_cmd,
+                                         stdout=subprocess.PIPE,
+                                         stderr=subprocess.STDOUT,
+                                         shell=True,
+                                         executable="/bin/bash")
+
+            env_dict = {}
+            output = module_p.communicate()[0]
+
+            # Loop over each environment variable key=value byte string
+            for entry in output.strip(b'\0').split(b'\0'):
+                # Split variable name and value
+                parts = entry.split(b'=', 1)
+                if len(parts) != 2:
+                    continue
+
+                # We'd really like to just pass byte strings to os.environ,
```

Comment:
https://docs.python.org/3/library/os.html#os.environb

Reply:
the more you know...